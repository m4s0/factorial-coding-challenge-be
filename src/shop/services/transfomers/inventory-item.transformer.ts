import { InventoryItem } from '@Shop/entities/inventory-item.entity';
import { InventoryItemOutput } from '@Shop/types/inventory-item.output';
import { ProductOption } from '@Shop/entities/product-option.entity';
import { InventoryStatusForProduct } from '@Shop/types/inventory-status-for-product';
import { ProductOptionOutput } from '@Shop/types/product-option.output';
import { OptionRule } from '@Shop/entities/option-rule.entity';
import { OptionRuleOutput } from '@Shop/types/option-rule.output';
import { OptionPriceRule } from '@Shop/entities/option-price-rule.entity';
import { OptionPriceRuleOutput } from '@Shop/types/option-price-rule.output';

function transformOptionPriceRule(
  optionPriceRule: OptionPriceRule,
): OptionPriceRuleOutput {
  return {
    id: optionPriceRule.id,
    price: Number(optionPriceRule.price.toFixed(2)),
    targetOptionId: optionPriceRule.targetOptionId,
    dependentOptionId: optionPriceRule.dependentOptionId,
    isActive: optionPriceRule.isActive,
    createdAt: optionPriceRule.createdAt.toISOString(),
    updatedAt: optionPriceRule.updatedAt.toISOString(),
  };
}

function transformOptionRule(optionRule: OptionRule): OptionRuleOutput {
  return {
    id: optionRule.id,
    ruleType: optionRule.ruleType,
    ifOptionId: optionRule.ifOptionId,
    thenOptionId: optionRule.thenOptionId,
    isActive: optionRule.isActive,
    createdAt: optionRule.createdAt.toISOString(),
    updatedAt: optionRule.updatedAt.toISOString(),
  };
}

function transformProductOption(
  productOption: ProductOption,
  optionInventory?: InventoryStatusForProduct[],
): ProductOptionOutput {
  return {
    id: productOption.id,
    name: productOption.name,
    displayName: productOption.displayName,
    basePrice: Number(Number(productOption.basePrice).toFixed(2)),
    isActive: productOption.isActive,
    rulesAsCondition: productOption.rulesAsCondition.map((optionRule) => {
      return transformOptionRule(optionRule);
    }),
    rulesAsResult: productOption.rulesAsResult.map((optionRule) => {
      return transformOptionRule(optionRule);
    }),
    priceRules: productOption.priceRules.map((optionPriceRule) => {
      return transformOptionPriceRule(optionPriceRule);
    }),
    createdAt: productOption.createdAt.toISOString(),
    updatedAt: productOption.updatedAt.toISOString(),
    ...(optionInventory && {
      inStock:
        optionInventory.find((inv) => inv.productOptionId === productOption.id)
          ?.inStock || false,
    }),
  };
}

export function transformInventoryItem(
  inventoryItem: InventoryItem,
): InventoryItemOutput {
  return {
    id: inventoryItem.id,
    quantity: inventoryItem.quantity,
    outOfStock: inventoryItem.outOfStock,
    productOptionId: inventoryItem.productOptionId,
    ...(inventoryItem.productOption && {
      productOption: transformProductOption(inventoryItem.productOption),
    }),
    createdAt: inventoryItem.createdAt.toISOString(),
    updatedAt: inventoryItem.updatedAt.toISOString(),
  };
}
