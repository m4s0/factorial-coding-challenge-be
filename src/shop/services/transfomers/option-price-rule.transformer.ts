import { OptionPriceRule } from '@Shop/entities/option-price-rule.entity';
import { OptionPriceRuleOutput } from '@Shop/types/option-price-rule.output';
import { transformProductOption } from '@Shop/services/transfomers/product-option.transform';

export function transformOptionPriceRule(
  optionPriceRule: OptionPriceRule,
): OptionPriceRuleOutput {
  return {
    id: optionPriceRule.id,
    price: Number(optionPriceRule.price),
    targetOptionId: optionPriceRule.targetOptionId,
    targetOption: transformProductOption(optionPriceRule.targetOption),
    dependentOptionId: optionPriceRule.dependentOptionId,
    isActive: optionPriceRule.isActive,
    createdAt: optionPriceRule.createdAt.toISOString(),
    updatedAt: optionPriceRule.updatedAt.toISOString(),
  };
}
