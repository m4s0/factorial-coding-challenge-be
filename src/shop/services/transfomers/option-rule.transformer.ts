import { OptionRule } from '@Shop/entities/option-rule.entity';
import { OptionRuleOutput } from '@Shop/types/option-rule.output';
import { transformProductOption } from '@Shop/services/transfomers/product-option.transform';

export function transformOptionRule(optionRule: OptionRule): OptionRuleOutput {
  return {
    id: optionRule.id,
    ruleType: optionRule.ruleType,
    ifOptionId: optionRule.ifOptionId,
    ...(optionRule.ifOption && {
      ifOption: transformProductOption(optionRule.ifOption),
    }),
    thenOptionId: optionRule.thenOptionId,
    ...(optionRule.thenOption && {
      thenOption: transformProductOption(optionRule.thenOption),
    }),
    isActive: optionRule.isActive,
    createdAt: optionRule.createdAt.toISOString(),
    updatedAt: optionRule.updatedAt.toISOString(),
  };
}
