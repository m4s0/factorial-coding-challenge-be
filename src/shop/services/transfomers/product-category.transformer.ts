import { Product } from '@Shop/entities/product.entity';
import { ProductOutput } from '@Shop/types/product.output';
import { transformProductOptionGroup } from '@Shop/services/transfomers/product-option-group.transformer';
import { ProductCategoryOutput } from '../../types/product-category.output';
import { ProductCategory } from '../../entities/product-category.entity';

function transformProduct(product: Product): ProductOutput {
  return {
    id: product.id,
    name: product.name,
    description: product.description,
    basePrice: Number(Number(product.basePrice).toFixed(2)),
    isActive: product.isActive,
    optionGroups: product.optionGroups.map((productOptionGroup) => {
      return transformProductOptionGroup(productOptionGroup);
    }),
    createdAt: product.createdAt.toISOString(),
    updatedAt: product.updatedAt.toISOString(),
  };
}

export function transformProductCategory(
  category: ProductCategory,
): ProductCategoryOutput {
  return {
    id: category.id,
    name: category.name,
    description: category.description,
    isActive: category.isActive,
    ...(category.products &&
      category.products.length > 0 && {
        products: category.products.map((product) => transformProduct(product)),
      }),
    createdAt: category.createdAt.toISOString(),
    updatedAt: category.updatedAt.toISOString(),
  };
}
