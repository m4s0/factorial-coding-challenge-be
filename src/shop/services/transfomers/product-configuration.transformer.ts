import { Product } from '@Shop/entities/product.entity';
import { ProductOptionGroup } from '@Shop/entities/product-option-group.entity';
import { OptionRule } from '@Shop/entities/option-rule.entity';
import { transformProduct } from '@Shop/services/transfomers/product.transformer';
import { transformOptionRule } from '@Shop/services/transfomers/option-rule.transformer';
import { transformProductOptionGroup } from '@Shop/services/transfomers/product-option-group.transformer';
import { ProductConfigurationOutput } from '@Shop/types/product-configuration.output';

export function transformProductConfiguration(
  product: Product,
  optionGroups: ProductOptionGroup[],
  rules: OptionRule[],
): ProductConfigurationOutput {
  return {
    product: transformProduct(product),
    optionGroups: optionGroups.map((optionGroup) =>
      transformProductOptionGroup(optionGroup),
    ),
    rules: rules.map((rule) => transformOptionRule(rule)),
  };
}
