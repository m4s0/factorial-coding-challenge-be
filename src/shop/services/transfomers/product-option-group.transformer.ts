import { transformProductOption } from '@Shop/services/transfomers/product-option.transform';
import { InventoryStatusForProduct } from '@Shop/types/inventory-status-for-product';
import { transformProduct } from '@Shop/services/transfomers/product.transformer';
import { ProductOptionGroup } from '../../entities/product-option-group.entity';
import { ProductOptionGroupOutput } from '../../types/product-option-group.output';

export function transformProductOptionGroup(
  productOptionGroup: ProductOptionGroup,
  optionInventory?: InventoryStatusForProduct[],
): ProductOptionGroupOutput {
  return {
    id: productOptionGroup.id,
    name: productOptionGroup.name,
    displayName: productOptionGroup.displayName,
    ...(productOptionGroup.options && {
      options: productOptionGroup.options.map((option) =>
        transformProductOption(option, optionInventory),
      ),
    }),
    ...(productOptionGroup.product && {
      product: transformProduct(productOptionGroup.product, optionInventory),
    }),
    createdAt: productOptionGroup.createdAt.toISOString(),
    updatedAt: productOptionGroup.updatedAt.toISOString(),
  };
}
