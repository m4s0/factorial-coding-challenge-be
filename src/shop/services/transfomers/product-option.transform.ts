import { ProductOptionOutput } from '@Shop/types/product-option.output';
import { ProductOption } from '@Shop/entities/product-option.entity';
import { transformInventoryItem } from '@Shop/services/transfomers/inventory-item.transformer';
import { transformOptionRule } from '@Shop/services/transfomers/option-rule.transformer';
import { transformOptionPriceRule } from '@Shop/services/transfomers/option-price-rule.transformer';
import { transformProductOptionGroup } from '@Shop/services/transfomers/product-option-group.transformer';
import { InventoryStatusForProduct } from '@Shop/types/inventory-status-for-product';

export function transformProductOption(
  option: ProductOption,
  optionInventory?: InventoryStatusForProduct[],
): ProductOptionOutput {
  return {
    id: option.id,
    name: option.name,
    displayName: option.displayName,
    basePrice: option.basePrice ? option.basePrice : null,
    isActive: option.isActive,
    optionGroupId: option.optionGroupId,
    ...(option.optionGroup && {
      optionGroup: transformProductOptionGroup(option.optionGroup),
    }),
    ...(option.inventoryItems && {
      inventoryItems: option.inventoryItems.map((item) =>
        transformInventoryItem(item),
      ),
    }),
    ...(option.rulesAsCondition && {
      rulesAsCondition: option.rulesAsCondition.map((rule) =>
        transformOptionRule(rule),
      ),
    }),
    ...(option.rulesAsResult && {
      rulesAsResult: option.rulesAsResult.map((rule) =>
        transformOptionRule(rule),
      ),
    }),
    ...(option.priceRules && {
      priceRules: option.priceRules.map((rule) =>
        transformOptionPriceRule(rule),
      ),
    }),
    ...(option.createdAt && {
      createdAt: option.createdAt.toISOString(),
    }),
    ...(option.updatedAt && {
      updatedAt: option.updatedAt.toISOString(),
    }),
    ...(optionInventory && {
      inStock:
        optionInventory.find((inv) => inv.productOptionId === option.id)
          ?.inStock || false,
    }),
  };
}
