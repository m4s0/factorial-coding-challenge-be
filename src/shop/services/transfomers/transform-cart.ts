import { Cart } from '@Shop/entities/cart.entity';
import { CartOutput } from '@Shop/types/cart.output';

export function transformCart(cart: Cart): CartOutput {
  return {
    id: cart.id,
    totalPrice: cart.totalPrice,
    createdAt: cart.createdAt.toISOString(),
    updatedAt: cart.updatedAt.toISOString(),
    items: cart.items.map((item) => ({
      id: item.id,
      productId: item.productId,
      quantity: item.quantity,
      price: item.price,
      totalPrice: item.totalPrice,
      createdAt: item.createdAt.toISOString(),
      updatedAt: item.updatedAt.toISOString(),
      product: {
        id: item.product.id,
        name: item.product.name,
        description: item.product.description,
        price: item.product.basePrice,
        categoryId: item.product.categoryId,
        createdAt: item.product.createdAt.toISOString(),
        updatedAt: item.product.updatedAt.toISOString(),
      },
      itemOptions: item.itemOptions.map((itemOption) => ({
        id: itemOption.id,
        createdAt: itemOption.createdAt.toISOString(),
        updatedAt: itemOption.updatedAt.toISOString(),
        optionId: itemOption.optionId,
        option: {
          id: itemOption.option.id,
          name: itemOption.option.name,
          displayName: itemOption.option.displayName,
          price: itemOption.price,
          optionGroupId: itemOption.option.optionGroupId,
          createdAt: itemOption.option.createdAt.toISOString(),
          updatedAt: itemOption.option.updatedAt.toISOString(),
        },
      })),
    })),
  };
}
