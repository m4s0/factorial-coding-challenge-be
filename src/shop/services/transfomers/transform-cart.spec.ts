import { transformCart } from '@Shop/services/transfomers/transform-cart';
import { Cart } from '@Shop/entities/cart.entity';
import { User } from '@User/entities/user.entity';
import { CartItem } from '@Shop/entities/cart-item.entity';
import { CartItemOption } from '@Shop/entities/cart-item-option.entity';
import { Product } from '@Shop/entities/product.entity';

describe('transformOptionGroup', () => {
  it('should transform cart correctly', () => {
    const cart = {
      id: '1fd7b09c-6da1-4f84-afe5-9ab0117c4cab',
      totalPrice: 835.0,
      user: {} as User,
      userId: '3e626cc9-9368-4b62-984b-bb94b1e58794',
      createdAt: new Date('2024-03-15T10:00:00Z'),
      updatedAt: new Date('2024-03-15T10:00:00Z'),
      items: [
        {
          id: 'a87f1507-55f5-41e1-b66e-34b245dd3e00',
          cart: {} as Cart,
          cartId: '1fd7b09c-6da1-4f84-afe5-9ab0117c4cab',
          productId: '09e51e85-e64e-4b40-a5d5-d791dfdcdd2a',
          quantity: 1,
          price: 835.0,
          totalPrice: 835.0,
          createdAt: new Date('2024-03-15T10:00:00Z'),
          updatedAt: new Date('2024-03-15T10:00:00Z'),
          product: {
            id: '09e51e85-e64e-4b40-a5d5-d791dfdcdd2a',
            name: 'Trailblazer Pro',
            description:
              'High-performance mountain bike for serious trail riders',
            basePrice: 800.0,
            categoryId: '3fee385c-8b66-4571-9d24-4c18e9223515',
            createdAt: new Date('2024-03-15T10:00:00Z'),
            updatedAt: new Date('2024-03-15T10:00:00Z'),
          } as Product,
          itemOptions: [
            {
              id: '1db5f36b-d590-4334-83db-0004701f8965',
              createdAt: new Date('2024-03-15T10:00:00Z'),
              updatedAt: new Date('2024-03-15T10:00:00Z'),
              optionId: '195aa790-a520-444b-95c2-f6c5f4cf63b1',
              option: {
                id: '195aa790-a520-444b-95c2-f6c5f4cf63b1',
                name: 'matte',
                displayName: 'Matte',
                basePrice: 35.0,
                optionGroupId: 'a95635bd-fd69-44b7-8f63-cab9fbaacbf1',
                createdAt: new Date('2024-03-15T10:00:00Z'),
                updatedAt: new Date('2024-03-15T10:00:00Z'),
              },
            } as CartItemOption,
          ],
        } as CartItem,
      ],
    };

    const result = transformCart(cart);

    expect(result).toMatchSnapshot();
  });
});
