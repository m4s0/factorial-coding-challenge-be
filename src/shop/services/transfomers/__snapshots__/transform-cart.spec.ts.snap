// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformOptionGroup should transform cart correctly 1`] = `
{
  "createdAt": "2024-03-15T10:00:00.000Z",
  "id": "1fd7b09c-6da1-4f84-afe5-9ab0117c4cab",
  "items": [
    {
      "createdAt": "2024-03-15T10:00:00.000Z",
      "id": "a87f1507-55f5-41e1-b66e-34b245dd3e00",
      "itemOptions": [
        {
          "createdAt": "2024-03-15T10:00:00.000Z",
          "id": "1db5f36b-d590-4334-83db-0004701f8965",
          "option": {
            "createdAt": "2024-03-15T10:00:00.000Z",
            "displayName": "Matte",
            "id": "195aa790-a520-444b-95c2-f6c5f4cf63b1",
            "name": "matte",
            "optionGroupId": "a95635bd-fd69-44b7-8f63-cab9fbaacbf1",
            "price": undefined,
            "updatedAt": "2024-03-15T10:00:00.000Z",
          },
          "optionId": "195aa790-a520-444b-95c2-f6c5f4cf63b1",
          "updatedAt": "2024-03-15T10:00:00.000Z",
        },
      ],
      "price": 835,
      "product": {
        "categoryId": "3fee385c-8b66-4571-9d24-4c18e9223515",
        "createdAt": "2024-03-15T10:00:00.000Z",
        "description": "High-performance mountain bike for serious trail riders",
        "id": "09e51e85-e64e-4b40-a5d5-d791dfdcdd2a",
        "name": "Trailblazer Pro",
        "price": 800,
        "updatedAt": "2024-03-15T10:00:00.000Z",
      },
      "productId": "09e51e85-e64e-4b40-a5d5-d791dfdcdd2a",
      "quantity": 1,
      "totalPrice": 835,
      "updatedAt": "2024-03-15T10:00:00.000Z",
    },
  ],
  "totalPrice": 835,
  "updatedAt": "2024-03-15T10:00:00.000Z",
}
`;
